app/controllers/pensions/overseas/incomeFromOverseasPension/PensionPaymentsController.scala

```diff
@@ -0,0 +1,124 @@
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendController
+import utils.{Clock, SessionHelper}
+import controllers.pensions.routes.PensionsSummaryController
+import controllers.pensions.overseas.incomeFromOverseasPension.routes.PensionPaymentsController
+import models.{AuthorisationRequest, User}
+import views.html.pensions.overseas.incomeFromOverseasPension.PensionPaymentsView
+
+import javax.inject.Inject
+import scala.concurrent.{ExecutionContext, Future}
+
+class PensionPaymentsController @Inject()(authAction: AuthorisedAction,
+                                          pensionPaymentsView: PensionPaymentsView,
+                                          pensionSessionService: PensionSessionService,
+                                          errorHandler: ErrorHandler)
+                                         (implicit mcc: MessagesControllerComponents, appConfig: AppConfig, ec: ExecutionContext, clock: Clock)
+  extends FrontendController(mcc) with I18nSupport with SessionHelper with FormUtils {
+
+  def amountForm(user: User): Form[(Option[BigDecimal], Option[BigDecimal])] = OptionalTupleAmountForm.amountForm(
+    emptyFieldKey1 = "overseasPension.pensionPayments.amountBeforeTax.noEntry",
+    wrongFormatKey1 = s"overseasPension.pensionPayments.amountBeforeTax.incorrectFormat.${if (user.isAgent) "agent" else "individual"}",
+    exceedsMaxAmountKey1 = "overseasPension.pensionPayments.amountBeforeTax.tooBig",
+    emptyFieldKey2 = "common.pensions.error.amount.noEntry",
+    wrongFormatKey2 = "overseasPension.pensionPayments.nonUkTaxPaid.incorrectFormat",
+    exceedsMaxAmountKey2 = "common.pensions.error.amount.overMaximum"
+  )
+
+  def show(taxYear: Int, index: Option[Int]): Action[AnyContent] = (authAction andThen taxYearAction(taxYear)).async { implicit request =>
+    pensionSessionService.getPensionSessionData(taxYear, request.user).map {
+      case Right(Some(data)) =>
+        validateIndex(index, data.pensions.incomeFromOverseasPensions.overseasIncomePensionSchemes) match {
+          case Some(i) => populateForm(data, taxYear, i)
+          case None => Redirect(PensionsSummaryController.show(taxYear)) // Todo should redirect to another page
+        }
+      case _ =>
+        //TODO: - Redirect to Overseas Pensions cya page??
+        Redirect(PensionsSummaryController.show(taxYear))
+    }
+  }
+
+
+  def submit(taxYear: Int, index: Option[Int]): Action[AnyContent] = authAction.async { implicit request =>
+
+    pensionSessionService.getPensionSessionData(taxYear, request.user).flatMap {
+      case Right(Some(data)) =>
+        validateIndex(index, data.pensions.incomeFromOverseasPensions.overseasIncomePensionSchemes) match {
+          case Some(i) => amountForm(request.user).bindFromRequest.fold(
+            formWithErrors => Future.successful(BadRequest(pensionPaymentsView(formWithErrors, taxYear, i))),
+            amounts =>
+              updatePensionScheme(data, amounts, taxYear, i))
```

Comment:
I would have unpacked the tuple here, giving:

```
updatePensionScheme(data, amounts._1, amounts._2, taxYear, i))
```

And:

```
  private def updatePensionScheme(data: PensionsUserData, amountBeforeTaxOpt: Option[BigDecimal], nonUkTaxPaidOpt: Option[BigDecimal], taxYear: Int, index: Int)
```

Seems clearer to me.

Reply:
Fixed.

Reply:
Confirmed.